lua/mini/misc.lua

```diff
@@ -409,6 +409,102 @@ MiniMisc.zoom = function(buf_id, config)
   end
 end
 
+--- Restore the cursor on file reopen
+---
+--- What it does:
+--- - When reopening a file, this will make sure the cursor is placed back to
+---   the position where you left before.
+--- - This implements :help restore-cursor in a nicer way.
+--- - Creates autocommand for |BufRead| and |FileType| event with
+---   |MiniMisc.restore_cursor|
+---
+--- If you want to lazy load this use the |BufReadPre| event.
+---
+---@param opts table Allows to set options for |MiniMisc.restore_cursor|
+---   You can set the two options, `ignore_buftype` and `ignore_filetype` as
+---   an array of strings.
+---   Example:
+---   ```
+---   {
+---     ignore_buftype = { "quickfix", "nofile", "help" },
+---     ignore_filetype = { "gitcommit", "gitrebase" },
+---   }
+---   ```
+---   Default: `{}`.
+---
+---@usage >
+---   require('mini.misc').setup_restore_cursor()
+MiniMisc.setup_restore_cursor = function(opts)
```

Comment:
Let's follow this as general design:
- `setup_restore_cursor()`:
    - Receives `opts` with at least two possible fields `ignore_buftype` and `ignore_filetype`. Their defaults are as in example (not empty). Use the following approach:
     ```lua
    if not H.is_array_of(x, H.is_string) then
      H.error('In `setup_restore_cursor()` `opts.x` should be an array of strings.')
    end
    ``` 
    - Validates `opts` and ensures defaults are in place.
    - Creates Neovim<0.7 compatible autocommand following `:h restore-cursor`. I think something like this should work:
    ```lua
    local au_command = string.format(
      [[augroup MiniMiscRestoreCursor
          au!
          au BufRead * au FileType <buffer> ++once lua MiniMisc.restore_cursor(%s)
        augroup END]],
      vim.inspect(opts)
    )
    vim.api.nvim_exec(au_command, false)
    ```
- `restore_cursor()`:
    - Also takes `opts` as described earlier, but doesn't validate it.
    - Makes all necessary checks that prohibit restoring cursor.
    - Restores cursor with `vim.cmd([[normal! g'"zvzz]])`.

Reply:
The au_command doesn't work:

`misc.lua:465: Vim:E492: Not an editor command:   ignore_buftype = { "quickfix", "nofile", "help" },`

Reply:
> The au_command doesn't work:
> 
> `misc.lua:465: Vim:E492: Not an editor command: ignore_buftype = { "quickfix", "nofile", "help" },`

Ah, yes. Try replacing `vim.inspect(opts)` with `vim.inspect(opts, { newline = ' ', indent = '' })`.

Reply:
Interesting...

Still, current logic of triggering of every `BufRead` and `FileType` is not ideal. For ignored filetypes it means moving cursor twice (to restore and then back to first line).

Further investigation why documented solution doesn't work (when on surface it should) seems like the best course of action here. It can work in Vim (from where it is ported) but not in Neovim, or in both.

Reply:
It works fine if you use `nvim_create_autocmd`.

Reply:
> It works fine if you use `nvim_create_autocmd`.

With current implementation (triggers on both `BufRead` and `FileType`) or if rewritten to create a `FileType` autocommand on `BufRead`? If former, then see previous comment on it being not ideal. If latter, then it is some weird behavior.

Reply:
I've fixed it.